src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java

```diff
@@ -0,0 +1,235 @@
+
+	protected ByteBuffer	locations;
+	protected IntBuffer		locations2;
+	protected ByteBuffer	timestamps;
+	protected IntBuffer		timestamps2;
+	protected ShortBuffer	sectorCounts;
+	protected List<Boolean>	sectorUsed;
+
+	/**
+	 * Create a new RegionFile object representing the region file at the given path and load it into memory.
+	 * 
+	 * @throws IllegalArgumentException
+	 *             if the file is smaller than 4kiB
+	 * @throws NoSuchFileException
+	 *             if the file does not exist
+	 * @see #open(Path)
+	 * @author piegames
+	 */
+	public RegionFile(Path file) throws IOException {
+		this.file = Objects.requireNonNull(file);
+
+		if (!Files.exists(file))
+			throw new NoSuchFileException(file.toString());
+		if (Files.size(file) < 4096 * 2)
+			throw new IllegalArgumentException("File size must be at least 4kiB, is this file corrupt?");
+		FileChannel raf = FileChannel.open(file, StandardOpenOption.READ);
+
+		locations = ByteBuffer.allocate(4096);
+		raf.read(locations);
+		locations.flip();
+		locations2 = locations.asIntBuffer();
+
+		timestamps = ByteBuffer.allocate(4096);
+		raf.read(timestamps);
+		timestamps.flip();
+		timestamps2 = timestamps.asIntBuffer();
+
+		chunks = new Chunk[32 * 32];
+		chunksUpToDate = new boolean[32 * 32];
+		for (int i = 0; i < 1024; i++) {
+			int chunkPos = locations2.get(i) >>> 8;
+			int chunkLength = locations2.get(i) & 0xFF;
+			if (chunkPos > 0) {
+				/* i & 31 retrieves the last 5 bit which store the x coordinate */
+				chunks[i] = new Chunk(i & 31, i >> 5, timestamps2.get(i), raf, chunkPos, chunkLength);
+			} else
+				chunks[i] = null;
+			chunksUpToDate[i] = true;
+		}
+	}
+
+	/**
+	 * Returns the {@link Chunk} at the given coordinate
+	 *
+	 * @param x
+	 *            the x coordinate relative to this region file's origin
+	 * @param z
+	 *            the z coordinate relative to this region file's origin
+	 * @return the chunk at that coordinate or {@code null} if the chunk does not exist
+	 * @author piegames
+	 */
+	public Chunk getChunk(int x, int z) {
+		return chunks[(x & 31) | (z << 5)];
+	}
+
+	public Chunk getChunk(int i) {
+		return chunks[i];
+	}
+
+	/**
+	 * Set a chunk at the given coordinate
+	 * 
+	 * @param x
+	 *            the x coordinate relative to this region file's origin
+	 * @param z
+	 *            the z coordinate relative to this region file's origin
+	 * @param chunk
+	 *            The chunk to set. Use {@code null} to remove it.
+	 * @author piegames
+	 */
+	public synchronized void setChunk(int x, int z, Chunk chunk) {
+		setChunk((x & 31) | (z << 5), chunk);
+	}
+
+	public synchronized void setChunk(int i, Chunk chunk) {
+		chunks[i] = chunk;
+		chunksUpToDate[i] = false;
+	}
+
+	/**
+	 * Write all chunks that got changed to disk, update the file's header (chunk locations and timestamps) and truncate the file at the end
+	 * 
+	 * @author piegames
+	 */
+	public synchronized void writeChanges() throws IOException {
```

Comment:
I would just write chunks directly in setChunk(), and remove this method.
Some of this code can be copy n' pasted into setChunk() but if there is anything we can do to make it a little easer to read that would be nice.

Reply:
I initially thought of this, but if there are frequent changes (or say I modify each chunk a bit), there will be a lot of redundant disk writes and pointless position calculations.

Reply:
Let the user handle that. If there are not frequent changes you are just creating a cache invalidation hell for the user for no reason.
If the user wants to cache chunks, they know the best way to do that.
```
/*not final*/ Chunk cachedChunk = RegionFile#getChunk(x, z);
```
Lets make some changes to to the chunk
```
chunk = makeSomeChanges(chunk);
```
hmmm... another frequent change
```
chunk = anotherFrequentChange(chunk);
```
Now lets save all the changes at once
```
RegionFile#setChunk(x, z, chunk)
```

If you are not convinced yet, think about this...
What happens if you want to cache chunks on the world level instead of at the region level (like what Minecraft does internally)?

Caching the chunks would not be so bad if Chunk was immutable as it would hide the caching form the user. But it would still be a lot of useless (in my opinion) code that would just be there for the off change that someone spams getChunk(). And even if they do, it still might not be a problem (most modern OS's cache hot files anyway).